pyrolite/mineral/normative.py

```diff
@@ -301,308 +302,661 @@ def CIPW_norm(df, Fe_correction=None, adjust_all=False):
+                     df['Fe2O3'] + df['FeO'] / df['MgO'])
+
+    df['SI'] = 100 * df['MgO'] / (df['MgO'] + df['FeO'] + df['Fe2O3'] + df['Na2O'] + df['K2O'])
+
+    df['AR_True'] = (df['Al2O3'] + df['CaO'] + df['Na2O'] + df['K2O']) / (
+                df['Al2O3'] + df['CaO'] - df['Na2O'] - df['K2O'])
+    df['AR_False'] = (df['Al2O3'] + df['CaO'] + 2 * df['Na2O']) / (df['Al2O3'] + df['CaO'] - 2 * df['Na2O'])
+
+    df['AR'] = np.where(((df['K2O'] / df['Na2O']) >= 1) & ((df['K2O'] / df['Na2O']) <= 2.5) & (df['SiO2'] > 0.5),
+                        df['AR_True'], df['AR_False'])
+
+    df['Mg#'] = 100 * df['MgO'] / (df['MgO'] + df['FeO'])
+
+    # Mole Calculations
+    for oxide in majors + trace_oxides:
+        df['n_' + oxide] = df[oxide] / pt.formula(oxide).mass
+
+    # Combine minor oxides
+    df['n_FeO_corr'] = df['n_FeO'] + df['n_MnO'] + df['n_NiO'] + df['n_CoO']
+    df['n_CaO_corr'] = df['n_CaO'] + df['n_SrO'] + df['n_BaO']
+    df['n_K2O_corr'] = df['n_K2O'] + df['n_Rb2O'] + df['n_Cs2O']
+    df['n_Na2O_corr'] = df['n_Na2O'] + df['n_Li2O']
+    df['n_Cr2O3_corr'] = df['n_Cr2O3'] + df['n_V2O3']
+
+    # Corrected oxide molecular weight computations
+    df['x_MnO'] = df['n_MnO'] / df['n_FeO_corr']
+    df['x_FeO'] = df['n_FeO'] / df['n_FeO_corr']
+
+    df['x_NiO'] = df['n_NiO'] / df['n_FeO_corr']
+    df['x_CoO'] = df['n_CoO'] / df['n_FeO_corr']
+
+    df['x_SrO'] = df['n_SrO'] / df['n_CaO_corr']
+    df['x_BaO'] = df['n_BaO'] / df['n_CaO_corr']
+    df['x_CaO'] = df['n_CaO'] / df['n_CaO_corr']
+
+    df['x_Rb2O'] = df['n_Rb2O'] / df['n_K2O_corr']
+    df['x_Cs2O'] = df['n_Cs2O'] / df['n_K2O_corr']
+    df['x_K2O'] = df['n_K2O'] / df['n_K2O_corr']
+
+    df['x_Li2O'] = df['n_Li2O'] / df['n_Na2O_corr']
+    df['x_Na2O'] = df['n_Na2O'] / df['n_Na2O_corr']
+
+    df['x_V2O3'] = df['n_V2O3'] / df['n_Cr2O3_corr']
+    df['x_Cr2O3'] = df['n_Cr2O3'] / df['n_Cr2O3_corr']
+
+    df['n_FeO'] = df['n_FeO_corr']
+    df['n_CaO'] = df['n_CaO_corr']
+    df['n_K2O'] = df['n_K2O_corr']
+    df['n_Na2O'] = df['n_Na2O_corr']
+    df['n_Cr2O3'] = df['n_Cr2O3_corr']
+
+    # Corrected normative mineral molecular weight computations
+    def corr_m_wt(oxide):
+        return(df['x_'+ oxide] * pt.formula(oxide).mass)
+
+    df['MW_FeO_corr'] = corr_m_wt('MnO') + corr_m_wt('NiO') + corr_m_wt('CoO') + corr_m_wt('FeO')
+    df['MW_CaO_corr'] = corr_m_wt('BaO') + corr_m_wt('SrO') + corr_m_wt('CaO')
+    df['MW_K2O_corr'] = corr_m_wt('Rb2O') + corr_m_wt('Cs2O') + corr_m_wt('K2O')
+    df['MW_Na2O_corr'] = corr_m_wt('Li2O') + corr_m_wt('Na2O')
+    df['MW_Cr2O3_corr'] = corr_m_wt('V2O3') + corr_m_wt('Cr2O3')
+
+    # Corrected molecular weight of Ca, Na and Fe
+    df['MW_Ca_corr'] = df['MW_CaO_corr'] - element_AW['O']
+    df['MW_Na_corr'] = (df['MW_Na2O_corr'] - element_AW['O']) / 2
+    df['MW_Fe_corr'] = df['MW_FeO_corr'] - element_AW['O']
+
+    mineral_molecular_weights['Fe-Hy'] = df['MW_FeO_corr'] + 60.0843
```

Comment:
I think these 'corrected' mineral molecular weights are fairly key, particularly when dealing with trace elements I think. Not sure if there would be a more succinct way of doing this though?

Reply:
There could be, but it might take some thought. If we wrote out mineral compositions as combinations of oxides we could look them up and calculate adjusted molecular weights, e.g. for Mt:
```python
minerals = {
    "Mt": {"name": "magnetite", "formulae": "FeO Fe2O3", "mass": pd.formula("Fe3O4").mass}
    ...
}


def _update_molecular_masses(mineral_dict, corrected_mass_df):
    """
    Update a dictionary of mineral molecular masses based on their oxide
    components.

    Parameters
    ----------
    mineral_dict : :class:`dict`
        Dictionary of minerals containing compositions and molecular masses.
    corrected_mass_df : :class:`dict`
        Dataframe containing columns which include corrected molecular masses
        for specific oxide components.
    """
    for mineral, data in mineral_dict.items():
        composition = data["composition"]
        masses = 0.0

        for oxide in composition.split():
            # check for compositional components which have adjusted molecular weights
            if "MW_{}_corr".format(oxide) in corrected_mass_df:
                masses += corrected_mass_df["MW_{}_corr".format(oxide)]
            else:
                # get the components which don't have adjusted molecular weights
                masses += pf.formula(oxide)

        data["mass"] = masses

_update_molecular_masses(minerals, df)
```
It'd be a good idea to note somewhere here that the masses are 1D arrays here - just for understanding what's in this dictionary. Also - for functions directly and only related to this use case, it's a good idea to use the `_` prefix (generally used to denote private functions).

Reply:
i have implemented this with some changes to handle multiple counts of oxides in a mineral formula